src/LSST/system/Config.cpp

```diff
@@ -46,45 +46,119 @@ void Config::setup(std::string const& source) {
     _thisPtr = Ptr(new Config(source));
 }
 
-Config::Config(std::string const& source) {
-    if (source == "UNIT_TEST") {
-        // source ignored in this case.
-        _setValuesUnitTests();
-    } else {
-        throw invalid_argument("Config had invalid source " + source);
+void Config::reset() {
+    LCRITICAL("Config reseting global configuration!!!");
+    lock_guard<mutex> lock(_thisMtx);
+    _thisPtr.reset();
+}
+
+Config::Config(std::string const& source) : _source(source) {
+    try {
+        LINFO("Config trying to load yaml file ", _source);
+        _yaml = YAML::LoadFile(_source);
+    } catch (YAML::BadFile& ex) {
+        throw ConfigException(ERR_LOC, string("YAML::BadFile ") + ex.what());
+    }
+    verifyRequiredElements();
+}
+
+void Config::verifyRequiredElements() {
+    try {
+        LINFO("Config::verifyRequiredElements ", _source);
+        string str = getControlServerHost();
+        LINFO(" ControlServer:host=", str);
+
+        int i = getControlServerPort();
+        LINFO("ControlServer:port", i);
+
+        i = getControlServerThreads();
+        LINFO("ControlServer:threads", i);
+
+    } catch (exception const& ex) {
+        LCRITICAL("Config::verifyRequiredElements config file ", _source, " needs valid ", ex.what());
+        throw;
     }
 }
 
 Config& Config::get() {
     if (_thisPtr == nullptr) {
-        throw runtime_error("Config has not been setup.");
+        throw ConfigException(ERR_LOC, "Config has not been setup.");
     }
     return *_thisPtr;
 }
 
-void Config::_setValuesUnitTests() {
-    _setValue("controlServer", "port", "12678");
-    _setValue("controlServer", "threads", "1");
+int Config::getControlServerPort() {
+    string section = "ControlServer";
+    string key = "port";
+    return getSectionKeyAsInt(section, key, 1, 65535);
+}
+
+int Config::getControlServerThreads() {
+    string section = "ControlServer";
+    string key = "threads";
+    return getSectionKeyAsInt(section, key, 1, 3000);
+}
+
+string Config::getControlServerHost() {
+    string section = "ControlServer";
+    string key = "host";
+    return getSectionKeyAsString(section, key);
 }
 
-void Config::_setValue(string const& section, string const& key, string const& val) {
-    string secKey = section + ":" + key;
-    auto iter = _map.find(secKey);
-    if (iter != _map.end()) {
-        LWARN("Config trying to reset ", secKey, " from ", iter->second, " to ", val);
+int Config::getSectionKeyAsInt(string const& section, string const& key) {
+    if (!_yaml[section][key]) {
+        throw ConfigException(ERR_LOC, string("Config") + section + ": " + key + " is missing");
+    }
+    try {
+        int val = _yaml[section][key].as<int>();
+        return val;
+    } catch (exception const& ex) {
+        throw ConfigException(ERR_LOC, string("Config") + section + ": " + key + " failed int " + ex.what());
+    }
+}
```

Comment:
I think the `int` is just an edge case of `double`. Is it possible to rewrite the code to be something like this?

```
int Config::getSectionKeyAsInt(string const& section, string const& key) {
    return int(getSectionKeyAsDouble(section, key));
}
```

Reply:
I disagree. An entry like "threads: 1.6" or "port: 5000.1" are meaningless and likely wrong. As such, they should be flagged so they can fixed.  I've added a unit test, and yaml-cpp's `as<int>` does indeed throw when converting a double to an int.

Reply:
Sure. Thanks!